gdpr/anonymizers/base.py

```diff
@@ -0,0 +1,169 @@
+from typing import Any, Iterable, List, Optional, Union
+
+from django.core.exceptions import ImproperlyConfigured
+from django.db.models import Model
+
+from gdpr.encryption import numerize_key
+
+
+class RelationAnonymizer:
+    """
+    Base class for Anonymizers defining special relations.
+    """
+
+    def get_related_objects(self, obj: Model) -> Iterable:
+        raise NotImplementedError
+
+    def get_related_model(self) -> Model:
+        raise NotImplementedError
+
+
+class FieldAnonymizer:
+    """
+    Field anonymizer's purpose is to anonymize model field according to defined rule.
+    """
+
+    ignore_empty_values: bool = True
+    empty_values: List[Any] = [None]
+    _encryption_key = None
+    is_reversible: bool = True
+
+    class IrreversibleAnonymizationException(Exception):
+        pass
+
+    def __init__(self, ignore_empty_values: bool = None, empty_values: Optional[List[Any]] = None):
+        """
+        Args:
+            ignore_empty_values: defines if empty value of a model will be ignored or should be anonymized too
+            empty_values: define list of values which are considered as empty
+        """
+        self._ignore_empty_values = ignore_empty_values if ignore_empty_values is not None else self.ignore_empty_values
+        self._empty_values = empty_values if empty_values is not None else self.empty_values
+
+    def get_is_reversible(self, obj=None) -> bool:
+        """This method allows for custom implementation."""
+        return self.is_reversible
+
+    def get_anonymized_value_from_obj(self, obj, name: str, encryption_key: Optional[str] = None):
+        value = getattr(obj, name)
+        if self._ignore_empty_values and value in self._empty_values:
+            return value
+        self._encryption_key = encryption_key
```

Comment:
tohle není hezké. Neměl bys nastavovat třídní proměnnou v metodě, která není init. Nejde to jinak? Třeba to předávat do get_anonymized_value

Reply:
já myslím, že to právě dělá proto, že to nejde, protože tu hodnotu při initu nemá. Pak se, ale nemá ukládat vůbec nebo to má být rozdělené na dvě třídy, jedna, která něco definuje a druhá která se vytvoří až v době provádění se všemi potřebnými parametry. Factory pattern